packages/cli/src/modules/insights/insights.collection.service.ts

```diff
@@ -0,0 +1,253 @@
+import { Service } from '@n8n/di';
+import { In } from '@n8n/typeorm';
+import { DateTime } from 'luxon';
+import { Logger } from 'n8n-core';
+import type { ExecutionLifecycleHooks } from 'n8n-core';
+import {
+	UnexpectedError,
+	type ExecutionStatus,
+	type IRun,
+	type WorkflowExecuteMode,
+} from 'n8n-workflow';
+
+import { SharedWorkflow } from '@/databases/entities/shared-workflow';
+import { SharedWorkflowRepository } from '@/databases/repositories/shared-workflow.repository';
+import { InsightsMetadata } from '@/modules/insights/database/entities/insights-metadata';
+import { InsightsRaw } from '@/modules/insights/database/entities/insights-raw';
+
+import { InsightsConfig } from './insights.config';
+
+const shouldSkipStatus: Record<ExecutionStatus, boolean> = {
+	success: false,
+	crashed: false,
+	error: false,
+
+	canceled: true,
+	new: true,
+	running: true,
+	unknown: true,
+	waiting: true,
+};
+
+const shouldSkipMode: Record<WorkflowExecuteMode, boolean> = {
+	cli: false,
+	error: false,
+	retry: false,
+	trigger: false,
+	webhook: false,
+	evaluation: false,
+
+	// sub workflows
+	integrated: true,
+
+	// error workflows
+	internal: true,
+
+	manual: true,
+};
+
+type BufferedInsight = Pick<InsightsRaw, 'type' | 'value' | 'timestamp'> & {
+	workflowId: string;
+	workflowName: string;
+};
+
+@Service()
+export class InsightsCollectionService {
+	private readonly cachedMetadata: Map<string, InsightsMetadata> = new Map();
+
+	private bufferedInsights: Set<BufferedInsight> = new Set();
+
+	private flushInsightsRawBufferTimer: NodeJS.Timeout | undefined;
+
+	private isAsynchronouslySavingInsights = true;
+
+	private flushesInProgress: Set<Promise<void>> = new Set();
+
+	constructor(
+		private readonly sharedWorkflowRepository: SharedWorkflowRepository,
+		private readonly insightsConfig: InsightsConfig,
+		private readonly logger: Logger,
+	) {
+		this.logger = this.logger.scoped('insights');
+	}
+
+	startFlushingTimer() {
+		this.isAsynchronouslySavingInsights = true;
+		this.stopFlushingTimer();
+		this.flushInsightsRawBufferTimer = setTimeout(
+			async () => await this.flushEvents(),
+			this.insightsConfig.flushIntervalSeconds * 1000,
+		);
+	}
+
+	stopFlushingTimer() {
+		if (this.flushInsightsRawBufferTimer !== undefined) {
+			clearTimeout(this.flushInsightsRawBufferTimer);
+			this.flushInsightsRawBufferTimer = undefined;
+		}
+	}
+
+	async shutdown() {
```

Comment:
At first I thought this needed `@OnShutdown`, I take it if someone adds it by mistake it won't matter much.